### PR TITLE
Add topographic error calculation for hexagonal grid

### PR DESCRIPTION
**Pull request by TharindaDilshan**
This PR adds the functionality for Topographic Error calculation, computed by finding the first-best-matching and second-best-matching neurons in the hexagonal grid. 

Screenshot 2022-04-12 005139

The topographic error calculation is based on the above equation, which considers if the first-best-matching and second-best-matching neurons are neighbors in the SOM grid.

### CODE REVIEW

**Comment by JustGlowing** [comment]
Thanks for your submission, please have a look at guidelines to contribute 

https://github.com/JustGlowing/minisom#guidelines-to-contribute

**Comment by TharindaDilshan** [comment]
Updated

**Comment by JustGlowing** [comment]
I can't promise a quick review, but I'll try to do it as soon as I can.

File: minisom.py [hunk 1]
```diff
@@ -538,6 +537,28 @@ def topographic_error(self, data):
...
+                        self.get_euclidean_coordinates_from_index(bmu[1])] 
+                            for bmu in b2mu_inds]
+        b2mu_coords = array(b2mu_coords)
+        b2mu_neighbors = [(bmu1 >= bmu2-1) & ((bmu1 <= bmu2+1)) 
+                            for bmu1, bmu2 in b2mu_coords]
+        b2mu_neighbors = [neighbors.prod() for neighbors in b2mu_neighbors]
+        te = 1 - mean(b2mu_neighbors)
+        return te
+        
+    def get_euclidean_coordinates_from_index(self, index):
```
Comment (JustGlowing) [inline minisom.py:554]: In this repository private methods start with an underscore.

File: minisom.py [hunk 2]
```diff
@@ -538,6 +537,28 @@ def topographic_error(self, data):
         dxdy = hstack([diff(b2mu_x), diff(b2mu_y)])
         distance = norm(dxdy, axis=1)
         return (distance > t).mean()
+    
+    def topographic_error_hexagonal(self, data):
```
Comment (JustGlowing) [inline minisom.py:541]: If we have `topographic_error_hexagonal` we should also have `topographic_error_rectangular`. The naming convention about private methods applies also here.

File: minisom.py (hunk 1 above)
Comment (TharindaDilshan) [reply minisom.py]: Updated

File: minisom.py (hunk 2 above)
Comment (TharindaDilshan) [reply minisom.py]: Updated

**Comment by JustGlowing** [comment]
Just a heads up, everything seems fine. I'll run some tests once I get the chance and merge the PR.

**Comment by JustGlowing** [comment]
All the tests pass and the behaviour of the function seem okay. Please fix the following stylistic issues and your PR will be merged: 

```
$ pycodestyle minisom.py
minisom.py:533:1: W293 blank line contains whitespace
minisom.py:537:76: W291 trailing whitespace
minisom.py:538:76: W291 trailing whitespace
minisom.py:539:29: E127 continuation line over-indented for visual indent
minisom.py:541:64: W291 trailing whitespace
minisom.py:542:29: E127 continuation line over-indented for visual indent
minisom.py:546:1: W293 blank line contains whitespace
minisom.py:557:1: W293 blank line contains whitespace
minisom.py:564:54: E228 missing whitespace around modulo operator
```

**Comment by TharindaDilshan** [comment]
Updated

**Comment by JustGlowing** [comment]
I still see an issue

```
$ pycodestyle minisom.py
minisom.py:539:25: E127 continuation line over-indented for visual indent
```

**Comment by JustGlowing** [comment]
just merged, congrats 🥳

**Pull request merged**